services/api/src/resources/billing/resolvers.ts

```diff
@@ -0,0 +1,136 @@
+import * as R from 'ramda';
+import billingModel, { AddBillingModifierInput, BillingModifier, BillingModifiersInput, UpdateBillingModifierInput, DeleteBillingModifierInput, DeleteAllBillingGroupModifiersInput } from '../../models/billing';
+
+
+/**
+ * Create/Add Billing Modifier
+ *
+ * @param {obj} root The rootValue passed from the Apollo server configuration.
+ * @param {AddBillingModifierInput} args
+ * @param {ExpressContext} context this includes the context passed from the apolloServer query
+ *     { sqlClient, hasPermissions, keycloakGrant, requestCache }
+ *
+ * @return {BillingModifier} The created modifier
+ */
+type AddBillingModifierAlias = (root: any, args: AddBillingModifierInput, context: {models: any, hasPermission: any}) => Promise<BillingModifier>;
+export const addBillingModifier: AddBillingModifierAlias = async (root, args, context) => {
+  const { hasPermission } = context;
+  const { input } = args;
+
+  // Input Validation
+  if (R.isEmpty(input.group)) {
+    throw new Error('You must provide a billing group name or id');
+  }
+
+  if (R.isEmpty(R.omit(['group', 'startDate', 'endDate'], input))) {
+    throw new Error('You must provide a discount value or extra cost.');
+  }
+
+  if( new Date(input.startDate) >= new Date(input.endDate)) {
+    throw new Error('You must provide a start date before the end date.');
+  }
+
+  // Permissions
+  await hasPermission('group', 'add');
+
+  // Action
+  return billingModel.addBillingModifier(input);
+};
+
+/**
+ * Update Billing Modifier
+ *
+ * @param {obj} root The rootValue passed from the Apollo server configuration.
+ * @param {UpdateBillingModifierInput} args
+ * @param {ExpressContext} context this includes the context passed from the apolloServer query
+ *     { sqlClient, hasPermissions, keycloakGrant, requestCache }
+ *
+ * @return {String} Success
+ */
+export const updateBillingModifier = async (_, args: UpdateBillingModifierInput, context: {models: any, hasPermission: any}) => {
+  const { hasPermission } = context;
+  const { input : { id, patch } } = args;
+
+  // Permissions
+  await hasPermission('group', 'update');
```

Comment:
For most permissions checks, keycloak needs to know more context (attributes). In this case, it needs to know what group is being checked, eg https://github.com/amazeeio/lagoon/blob/master/services/api/src/resources/group/resolvers.ts#L159

Reply:
Addressed in: bcb60674748dc28ef2fa5f75027f9c618e78950c